File: exemple/ex02_hyperstatique.py
"""
ex02_hyperstatique.py — Poutre continue HYPERSTATIQUE (3 travées)
=================================================================
Cas : pont-route 3 travées (12 m + 16 m + 12 m), section rectangulaire
      variée (travée centrale plus inerte).
Objectif : calculer les lignes d'influence complètes et afficher les
           moments sur appuis + enveloppe du moment fléchissant.
"""

import matplotlib.pyplot as plt
import Tsaraloha.LIPoutreContinue as lipc

# ── 1. Géométrie et matériaux ──────────────────────────────────────────────────
L_spans = [12.0, 16.0, 12.0]                  # Longueurs des travées [m]
E_val   = 35e9                                  # Béton HP 35 GPa
I_rive  = 0.35 * 0.70**3 / 12                  # Inertie travées de rive [m⁴]
I_cent  = 0.40 * 0.80**3 / 12                  # Inertie travée centrale [m⁴]
steps   = 0.5                                   # Pas 50 cm

E = [E_val] * len(L_spans)
I = [I_rive, I_cent, I_rive]

# ── 2. Calcul des lignes d'influence ──────────────────────────────────────────
poutre = lipc.Hyperstatique(E, I, L_spans, steps)

BM  = poutre.bending_moments()   # [travée][section][alpha] en kN·m
SM  = poutre.support_moment      # [appui][section]          en kN·m
X   = poutre.points_x_coordinates(poutre.span_node_positions)

print(f"Nombre de travées : {poutre.number_of_spans}")
print(f"Appuis intermédiaires cumulés : {poutre.L_spans}")
print(f"Coefficients de flexibilité a : {[f'{a:.2e}' for a in poutre.a_spans]}")

# ── 3. Tracé — moments sur appuis (section critique travée 0) ──────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))
fig.suptitle("Lignes d'influence — Poutre hyperstatique 3 travées", fontsize=13)

# Moment sur appui intermédiaire (appui 1)
ax1.plot(X, SM[1], "b-", linewidth=2, label="M appui 1")
ax1.axhline(0, color="k", linewidth=0.8, linestyle="--")
ax1.fill_between(X, SM[1], alpha=0.15, color="blue")
ax1.set_title("L.I. Moment sur appui intermédiaire [kN·m]")
ax1.set_xlabel("Position charge α [m]")
ax1.legend()
ax1.grid(True, alpha=0.3)

# Enveloppe du moment en mi-travée 0 (section centrale)
mid_sec = len(BM[0]) // 2
ax2.plot(X, BM[0][mid_sec], "r-", linewidth=2, label=f"M mi-travée 0 (sec. {mid_sec})")
ax2.axhline(0, color="k", linewidth=0.8, linestyle="--")
ax2.fill_between(X, BM[0][mid_sec], alpha=0.15, color="red")
ax2.set_title("L.I. Moment en mi-travée 0 [kN·m]")
ax2.set_xlabel("Position charge α [m]")
ax2.legend()
ax2.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

File: LIPoutreContinue/examples/ex02_hyperstatique.cpp
// =============================================================================
//  ex02_hyperstatique.cpp — Poutre continue HYPERSTATIQUE (3 travées)
// =============================================================================
//
//  Cas  : pont-route 3 travées (12 m + 16 m + 12 m), section rectangulaire
//         variée (travée centrale plus inerte).
//  Objet: calculer les lignes d'influence complètes, afficher les moments sur
//         appuis + moment max absolu sur la structure.
//
//  Build:
//      cmake -B build -DTSARALOHA_BUILD_EXAMPLES=ON
//      cmake --build build --config Release
//      ./build/LIPoutreContinue/examples/Release/ex02_hyperstatique  (Windows)
//      ./build/LIPoutreContinue/examples/ex02_hyperstatique          (Linux)
// =============================================================================

#include <LIPoutreContinue/StructuralAnalysis.h>

#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

int main()
{
    std::cout << std::fixed << std::setprecision(4);

    // ── 1. Geometrie et materiaux ────────────────────────────────────────────
    std::vector<double> L_spans{ 12.0, 16.0, 12.0 };   // Longueurs [m]
    const double E_val = 35e9;                           // Beton HP 35 GPa
    const double I_rive = 0.35 * 0.70 * 0.70 * 0.70 / 12.0;  // Travees de rive
    const double I_cent = 0.40 * 0.80 * 0.80 * 0.80 / 12.0;  // Travee centrale

    std::vector<double> E{ E_val, E_val, E_val };
    std::vector<double> I{ I_rive, I_cent, I_rive };
    double steps = 0.5;

    std::cout << "=== Exemple 2 - Poutre hyperstatique 3 travees ===\n"
              << "  Travees : 12 m + 16 m + 12 m\n"
              << "  E = " << E_val / 1e9 << " GPa\n"
              << "  I_rive = " << I_rive << " m4   I_cent = " << I_cent << " m4\n"
              << "  pas = " << steps << " m\n\n";

    // ── 2. Calcul des lignes d'influence ─────────────────────────────────────
    Hyperstatique beam(E, I, L_spans, steps);

    // BM[travee][section][alpha] — dimensions en mémoire
    auto BM  = beam.BendingMoments();
    auto SF  = beam.ShearForce();
    auto Def = beam.Deflection();
    auto Rot = beam.Rotation();
    auto X   = beam.pointsXCoordinates(beam.SpanNodePositions);

    std::cout << "Nombre de travees  : " << beam.number_of_spans << "\n"
              << "Nœuds X totaux     : " << X.size() << "\n"
              << "Longueur totale    : " << X.back() << " m\n\n";

    // ── 3. Moments sur appuis (SupportMoment[appui][alpha]) ──────────────────
    std::cout << "Moments sur appuis :\n";
    for (size_t k = 0; k < beam.SupportMoment.size(); ++k) {
        double mMax = 0.0;
        for (double v : beam.SupportMoment[k])
            if (std::abs(v) > std::abs(mMax)) mMax = v;
        std::cout << "  Appui " << k << "  : M_max = " << mMax << " kN.m\n";
    }
    std::cout << "\n";

    // ── 4. Maximum global du moment flechissant ───────────────────────────────
    Position3D maxBM = findMaxAbsoluteValue3D(BM);
    std::cout << "Maximum global |M| :\n"
              << "  Travee " << maxBM.i
              << "  section " << maxBM.j
              << "  alpha " << maxBM.k
              << "  -> M = " << maxBM.val << " kN.m\n\n";

    // ── 5. Maximum global de la fleche ───────────────────────────────────────
    Position3D maxDef = findMaxAbsoluteValue3D(Def);
    std::cout << "Maximum global |w| :\n"
              << "  Travee " << maxDef.i
              << "  section " << maxDef.j
              << "  alpha " << maxDef.k
              << "  -> w = " << maxDef.val << " m/kN\n\n";

    // ── 6. Affichage de la L.I. de M en mi-travee de la travee centrale ──────
    size_t span_mid = 1;                          // travee centrale (index 1)
    size_t sec_mid  = BM[span_mid].size() / 2;   // section centrale

    std::cout << "L.I. moment flechissant — mi-travee 1 (section " << sec_mid << ") :\n";
    std::cout << "  alpha [m]   M [kN.m/kN]\n";
    for (size_t k = 0; k < BM[span_mid][sec_mid].size(); k += 5) {
        std::cout << "  " << std::setw(8) << X[k]
                  << "   " << BM[span_mid][sec_mid][k] << "\n";
    }

    std::cout << "\nTermine.\n";
    return 0;
}
